PhysX collider asset tutorial. (#1281)
o3de/o3de.org: content/docs/learning-guide/tutorials/assets/physx-colliders.md

weight: 100
toc: true
---

**Open 3D Engine (O3DE)** has a very robust set of options for generating PhysX collider assets. You can customize how PhysX collider assets should be generated for a mesh asset in **Scene Settings**. The generated PhysX colliders are stored in `.pxmesh` product assets and can be added to a PhysX Collider component by selecting **PhysicsMesh** from in the component's **Shape** property.

There are three PhysX asset collider types that can be used in different simulation scenarios. This topic explains the benefits and limitations of the different collider types to help you understand when you might choose each type. This topic also covers the basics of generating PhysX asset colliders with Scene Settings.

## Entity behavior

Before learning about collider types, you should understand the three types of entity behavior that influence the collider asset type you might choose in a given scenario:

* **Static** - Static entities can be collided with, but don't move, and are not affected by PhysX collisions and forces. Static entities can use any collider type.

> * **Kinematic** - Kinematic entities have a **PhysX Rigid Body** component and have movement that is driven by script. Kinematic entities can be collided with, but are not affected by PhysX collisions or forces. Kinematic entities can only use primitive and convex colliders.

Author: fixed

* **Dynamic** - Dynamic entities have a **PhysX Rigid Body** component and simulated movement that results from PhysX collisions and forces. Dynamic entities can only use primitive and convex colliders.

## PhysX collider assets

Collider assets are generated when a source asset is processed and are based on input meshes you specify in Scene Settings. You can use a single input mesh, multiple input meshes, and even specify that multiple input meshes should be merged and welded into a single mesh before the collider asset is generated. The source asset is not altered when the collider asset is generated. The options to combine, clean, weld, and decompose input meshes are performed as part of the process of generating the collider asset, and are discarded when the process job completes.

Because physics simulation can be computationally expensive, it's important to understand the various collider types, their limitations, and the scenarios in which they can be used.

> ### Triangle Mesh

Author: It's a UI property capitalized in the interface.

> As the name implies, triangle mesh colliders are composed of triangle meshes. You can create a simplified mesh in a digital content creation (DCC) application specifically for the collider, or generate a triangle mesh collider based on the render mesh of the asset. Triangle mesh colliders can closely approximate a complex render mesh, but can only be used with static entities. Because triangle mesh colliders can contain many vertices, and the collider asset is not required to be convex, they generally incur a higher performance cost than primitive colliders.

Author: fixed

Triangle mesh colliders can have multiple physics material assignments. Because the collider is derived from the input mesh, one physics material can be applied to the collider per each material assignment. This is particularly useful for ground and floor colliders that transition between materials such as concrete and dirt, or wood planks and carpet.

> In the image below, a triangle mesh collider asset is offset to the right of the input mesh asset that generated the collider. Note that the collider asset (right) accurately represents the render mesh (left) and that the collider supports separate physics materials for the front and back faces (magenta wireframe) and side faces (teal wireframe) of the mesh.

Author: Changed all these examples to something like:

Colliders occupy the same physical space as the input mesh when they are generated. In the image below, a triangle mesh collider asset is offset to the right of the input mesh asset that generated the collider so that you can clearly view the collider.

{{< image-width "/images/learning-guide/tutorials/assets/triangle-mesh-collider-example.png" "700" "An example triangle mesh collider asset." >}}

Triangle mesh colliders are commonly used for immobile environment objects that have complex shapes such as tree trunks, floors, large stone and architectural features, and statues.

### Primitive

> Primitive colliders are composed of sphere, box, or capsule primitive shapes. Because the shapes are defined by simple properties such as **Radius**, **Height**, and **Width**, primitive colliders generally offer the best simulation performance. Primitive colliders can be automatically fit to the input mesh. However, depending on the complexity of the input mesh, areas of the collider might fall within or far outside of the input mesh surface. This can result in collisions that aren't visually accurate. Complex input meshes can be decomposed into smaller parts, and have primitive colliders fit to each part. Primitive colliders can be used with static, kinematic, or dynamic entities and can only have one physics material assignment.

Author: Turns out fit is a mid-western idiosyncrasy. Changing to fitted.

In the image below, a primitive collider asset is offset to the right of the input mesh asset that generated the collider. The collider asset (right) is a simple box that has been automatically oriented and scaled to encompass the render mesh (left). Note that one corner of the primitive collider extends far from the input mesh and is penetrating the ground plane. This demonstrates one of the drawbacks of using a primitive collider on a complex input mesh. In a scenario where this is a dynamic entity, the ground plane collision would push the asset upward and the asset would not come to rest standing upright because of the sharp corner at the bottom of the primitive collider.

{{< image-width "/images/learning-guide/tutorials/assets/primitive-collider-example.png" "700" "An example primitive collider asset." >}}

Primitive colliders are commonly used when the input mesh closely resembles one of the simple primitive shapes, and in scenarios where fast dynamic collision resolution is more important than visual accuracy.

### Convex

> Convex colliders are automatically generated convex hulls. In convex hulls, the shared edges where faces meet are *reflex* angles that are greater than 180°. The convex hull is constructed within a limited number of vertices and is fit to the input mesh. Convex colliders can better approximate a complex input mesh than primitive colliders, but incur a greater performance cost. To a lesser extent than primitive colliders, convex colliders might also have areas where the collider surface falls inside or outside of the input mesh, which might result in collisions that aren't visually accurate. Complex input meshes can be decomposed into smaller parts, and have convex colliders fit to each part. Convex colliders can be used with static, kinematic, or dynamic entities and can only have one physics material assignment.

Author: Using the colloquial description, fitted, and specifying "For PhysX, the convex hull is constructed within a limited number of vertices."

In the image below, a convex collider asset is offset to the right of the input mesh asset that generated the collider. Note that the collider asset (right) is has been automatically generated to encompass and roughly approximate the silhouette shape of the render mesh (left). The convex collider, however, does not include detail such as the hole in the middle of the input mesh.

{{< image-width "/images/learning-guide/tutorials/assets/convex-collider-example.png" "700" "An example convex collider asset." >}}

Convex colliders are commonly used on entities with complex render meshes that must be dynamically simulated, such as interactive props, and entities that must be decomposed into smaller parts to provide collisions with a higher level of visual accuracy.

### Summary

The table below summarizes the most import information to know about the available collider types:

| Type | Entity Behavior | Description | Limitations |
| - | - | - | - |

> | **Triangle mesh** | Static | A collider composed of triangles. Triangle mesh colliders can closely approximate complex input meshes and can have more than one physics material assignment.  | Can only be used with static entities. Can be more computationally expensive to simulate than other collider types. |

Author: fixed

{{< note >}}
If you are unfamiliar with Scene Settings, refer to the [Mesh Asset tutorial](../mesh-assets) and check the [Scene Settings PhysX Tab](/docs/user-guide/assets/scene-settings/physx-tab) topic for in-depth information on the options for generating PhysX collider assets.
{{< /note >}}

You can follow the tutorial below using any source asset that contains at least one mesh.

1. In **O3DE Editor**, locate your source asset in **Asset Browser**. If you don't have an asset of your own, you can type `fbx` into the search field at the top of Asset Browser and use one of the provided `.fbx` files such as `sphere.fbx`.

    ![ Search for a specific mesh asset in Asset Browser. ](/images/learning-guide/tutorials/assets/meshes-search-asset-browser.png)

    If your asset has already been processed, you might see a preview image of the asset, and a list of product assets below the `.fbx` source asset.

1. **Right-click** the `.fbx` source asset and choose **Edit settings...** from the context menu to open Scene Settings.

>     ![ Open Scene Settings from Asset Browser. ](/images/learning-guide/tutorials/assets/meshes-edit-settings.png)

Author: The image is in the mesh tutorial PR.

    ![ Scene Settings PhysX tab. ](/images/learning-guide/tutorials/assets/physx-scene-settings.png)

    In the image above, there is a single **PhysX mesh group**. Each PhysX mesh group produces a `pxmesh` product asset. You can create additional PhysX mesh groups for a source asset by choosing **Add another physxmesh**.
    
    The **Name mesh** property contains the name of the source asset. The `.pxmesh` product asset of this PhysX mesh group uses this string for its name.

1. The **Select meshes** property reads **0 of 1 selected**. Choose the {{< icon browse-edit-select-files.svg >}} file select button to select which meshes to include in the PhysX mesh group. Meshes in the list are denoted by a purple mesh icon, as in the example below. You can select more than one mesh here if multiple meshes are available in the asset. If you do select more than one mesh, you might also want to activate the **Merge Meshes** and **Weld Vertices** properties to ensure the input mesh is optimized.

    ![ Selecting a PhysX mesh. ](/images/learning-guide/tutorials/assets/select-physx-mesh.png)

{{< note >}}

> To automatically assign meshes in the source asset to a Physx mesh group, add the suffix `_phys` to the mesh node name in your DCC application. Any meshes that have `_phys` postfixed to their node name in the source asset are excluded from the default render mesh group and are automatically added to a single Physx mesh group in Scene Settings.

Author: This still works. I tested it. Works with `.gltf` and `.glb` as well.

9. With just a PhysX Collider component, this is a static entity. PhysX objects can collide with this entity, but the entity will not move or react to PhysX collisions or forces.

{{< note >}}
If you intend for the entity to be static, you should also activate the **Static** property in the **Transform** component of the entity. The **Static** property optimizes the entity, ensuring the best runtime performance for static entities.
{{< /note >}}

10. If you intend to create a kinematic or dynamic entity, you must add a **PhysX Rigid Body** component as well. With the entity selected, in Entity Inspector, choose **Add Component** and select PhysX Rigid Body from the component list. Ensure the **Static** property in the entity's Transform component is not active.

{{< note >}}
Adding the PhysX Rigid Body component makes the entity dynamic. If you choose the {{< icon simulate-physics.svg >}} simulation button, the entity will drop with gravity.

If you want to create a kinematic entity that is animated through script, activate the **Kinematic** property in the PhysX Rigid Body component.
{{< /note >}}

## Decompose input meshes

> Assets that are composed of multiple meshes, or assets that have complex meshes, might require similarly complex PhysX collider assets. This is particularly true for kinematic and dynamic entities where triangle mesh colliders can't be used. In these scenarios, the input mesh can be decomposed into approximately convex parts, and primitive colliders or convex colliders can be generated and automatically fit to each part.

Author: Thanks. "approximately" was in the original docs and it always struck me as odd. Will remove.

{{< note >}}

> Mesh decomposition doesn't alter the input mesh. Mesh decomposition is part of the process of generating and fitting collider assets. The decomposed mesh is discarded when the collider process job is completed. Mesh decomposition can only be used with primitive or convex colliders.

Reviewer: I'm not sure what is meant by "the decomposed mesh is discarded". I think the second sentence could be clearer if it emphasized that it's the output that can be primitive or convex.

Author: Is this better?

Assets that are composed of multiple meshes, or assets that have complex meshes, might require similarly complex PhysX collider assets. This is particularly true for kinematic and dynamic entities where triangle mesh colliders can't be used. In these scenarios, the input mesh can be decomposed into convex parts. Primitive colliders or convex colliders can be automatically generated and fitted to each part and processed as collider `.pxmesh` product assets.

Mesh decomposition in O3DE uses the [V-HACD Library](https://github.com/kmammou/v-hacd). Refer to the preceding link for more information and example images demonstrating decomposition of complex meshes.
{{< /note >}}

To use mesh decomposition, do the following:

> 1. In Scene Settings, in the PhysX mesh group, choose either `Primitive` or `Convex` as the **Export As** property. 

Reviewer 2: ```suggestion
1. In Scene Settings, under **PhysX Mesh group**, for **Export As**, choose either `Primitive` or `Convex`.
```
* Changed "in the PhysX mesh group" to "under **PhysX Mesh group** to change this to a UI reference that matches the formatting and capitalization of the string (per a previous screenshot in this topic).
* Re-ordered some of the procedure for sequential flow and standard procedure phrasing.

> 1. Activate the **Decompose Meshes** option.

Reviewer 2: ```suggestion
1. Enable **Decompose Meshes**.
```
* For consistency with my earlier edits, changed "activate" to "enable".
* To avoid confusion over whether to call this an "option", "property", or "parameter", I suggest dropping the noun at the end.

> Activating Decompose Meshes reveals many options you can use to fine tune mesh decomposition. For guidance on mesh decomposition options, refer to the [Decompose Meshes](/docs/user-guide/assets/scene-settings/physx-tab/#decompose-meshes) topic in the User Guide.

Reviewer 2: ```suggestion
Enabling Decompose Meshes reveals many options that you can use to fine-tune mesh decomposition. For guidance on mesh decomposition options, refer to the [Decompose meshes](/docs/user-guide/assets/scene-settings/physx-tab/#decompose-meshes) section of the **Scene Settings PhysX Tab** topic.
```

### Primitive decomposition

> When Decompose meshes is activated for primitive colliders, the input mesh is decomposed into approximately convex parts, and the best fitting primitive shapes are automatically selected and transformed to encompass each part. In the example below, the input mesh (left) is decomposed in to 6 parts (right). A primitive collider is automatically fit to each part. Most parts are fit with box primitive colliders, but one part, near the bottom of the asset, has a capsule primitive collider.

Reviewer: Again I'd suggest emphasizing that it's the output which is primitive. As it is, it could be confusing whether it's the input or output that is primitive. And the parts are precisely convex.

Author: I have reworded both this paragraph and the convex paragraph to try to clarify this point.

When Decompose meshes is activated for primitive colliders, the input mesh is decomposed into convex parts, the best fitting primitive shapes are automatically selected and transformed to encompass each part, and the primitives are processed as a collider `.pxmesh` product asset. In the example below, the input mesh (left) is decomposed in to 6 parts and a primitive collider is automatically fitted to each part (right). Most parts are fitted with box primitive colliders, but one part, near the bottom of the asset, has a capsule primitive collider.

> When Decompose meshes is activated for primitive colliders, the input mesh is decomposed into convex parts, the best fitting primitive shapes are automatically selected and transformed to encompass each part, and the primitives are processed as a collider `.pxmesh` product asset. In the example below, the input mesh (left) is decomposed in to 6 parts and a primitive collider is automatically fitted to each part (right). Most parts are fitted with box primitive colliders, but one part, near the bottom of the asset, has a capsule primitive collider.

Reviewer 2: ```suggestion
When you enable Decompose Meshes for primitive colliders, the input mesh decomposes into convex parts, O3DE automatically selects the best-fitting primitive shapes and transforms them to encompass each part, and O3DE processes primitives as a collider `.pxmesh` product asset. In the following image, the input mesh (left) decomposes into six parts, and O3DE automatically fits a primitive collider to each part (right). Most parts are fitted with box primitive colliders. However, one part near the bottom of the asset has a capsule primitive collider.
```
Misc. edits similar to previous edits in this topic. Be sure to confirm that O3DE is the correct subject for these attempts to reword passive voice construction.

{{< image-width "/images/learning-guide/tutorials/assets/primitive-decompose.png" "700" "An example of mesh decomposition with primitive collider assets." >}}

### Convex decomposition

> When Decompose meshes is activated for convex colliders, the input mesh is decomposed into convex parts, a convex hull is generated for each part, and the convex hulls are process as a collider `.pxmesh` product asset. In the example below, the input mesh (left) is decomposed in to 6 parts and convex hulls are generated for each part (right). The convex colliders provide a fairly accurate representation of the render mesh that might be sufficient in many scenarios.

Reviewer 2: ```suggestion
When you enable Decompose Meshes for convex colliders, the input mesh decomposes into convex parts, O3DE generates a convex hull for each part, and O3DE processes convex hulls as a collider `.pxmesh` product asset. In the following image, the input mesh (left) decomposes into six parts, and O3DE generates convex hulls for each part (right). The convex colliders provide a fairly accurate representation of the render mesh, which may be sufficient in many scenarios.
```
See previous comment. Also, changed "might" to "may" to add a bit of conviction to the claim.

{{< image-width "/images/learning-guide/tutorials/assets/convex-decompose.png" "700" "An example of mesh decomposition with convex collider assets." >}}

> In both of the examples above, the generated collider asset provides a more accurate representation of the render mesh than a single primitive or convex collider can. Notice, for example, that in both results, the hole in the middle of the logo is not blocked by the collider surface as it was in the initial example of the convex collider asset type.

Reviewer 2: ```suggestion
In both of the previous images, the generated collider asset provides a more accurate representation of the render mesh than a single primitive or convex collider can. For example, notice that in both results, the collider surface doesn't block the hole in the middle of the logo, unlike in the earlier image of the convex collider asset type.
```
